clients/drcachesim/tests/record_filter_unit_tests.cpp

```diff
@@ -94,7 +147,176 @@ get_basic_counts(const std::string &trace_dir)
+        { { TRACE_TYPE_WRITE, 4, { 0xaa80 } }, { true, true } },
+        // Unit header.
+        { { TRACE_TYPE_MARKER, TRACE_MARKER_TYPE_TIMESTAMP, { 0x9 } }, { false, true } },
+        { { TRACE_TYPE_MARKER, TRACE_MARKER_TYPE_CPU_ID, { 0xa } }, { false, true } },
+        { { TRACE_TYPE_WRITE, 4, { 0xaa90 } }, { false, false } },
+        { { TRACE_TYPE_ENCODING, 0, { 0 } }, { false, true } },
+
+        // Unit header.
+        { { TRACE_TYPE_MARKER, TRACE_MARKER_TYPE_TIMESTAMP, { 0xb } }, { true, true } },
+        { { TRACE_TYPE_MARKER, TRACE_MARKER_TYPE_CPU_ID, { 0xc } }, { true, true } },
+        { { TRACE_TYPE_MARKER, TRACE_MARKER_TYPE_FUNC_ID, { 0xd } }, { false, true } },
+        { { TRACE_TYPE_MARKER, TRACE_MARKER_TYPE_FUNC_ARG, { 0xe } }, { false, true } },
+        { { TRACE_TYPE_MARKER, TRACE_MARKER_TYPE_FUNC_RETADDR, { 0xf } },
+          { false, true } },
+        { { TRACE_TYPE_INSTR, 4, { 0xaa80 } }, { true, false } },
+        { { TRACE_TYPE_READ, 4, { 0xaaa0 } }, { false, false } },
+        // The following entry is part of the expected output, but not the input. We
+        // will skip it in the parallel_shard_filter() loop below.
+        { { TRACE_TYPE_MARKER, TRACE_MARKER_TYPE_FILTER_BOUNDARY, { 0 } },
+          { true, true } },
+        // Unit header.
+        // Since this timestamp is greater than the last_timestamp_us set below, all
+        // later entries will be output regardless of the configured filter.
+        { { TRACE_TYPE_MARKER, TRACE_MARKER_TYPE_TIMESTAMP, { 0xabcdef } },
+          { true, true } },
+        { { TRACE_TYPE_MARKER, TRACE_MARKER_TYPE_CPU_ID, { 0xa0 } }, { true, true } },
+        { { TRACE_TYPE_MARKER, TRACE_MARKER_TYPE_FUNC_ID, { 0xa1 } }, { true, true } },
+        // Trace shard footer.
+        { { TRACE_TYPE_FOOTER, 0, { 0xa2 } }, { true, true } }
+    };
+
+    for (int k = 0; k < 2; ++k) {
+        auto stream = std::unique_ptr<local_stream_t>(new local_stream_t());
+        // Construct record_filter_func_ts.
+        std::vector<dynamorio::drmemtrace::record_filter_t::record_filter_func_t *>
+            filters;
+        auto cache_filter =
+            std::unique_ptr<dynamorio::drmemtrace::record_filter_t::record_filter_func_t>(
+                new dynamorio::drmemtrace::cache_filter_t(
+                    /*cache_associativity=*/1, /*cache_line_size=*/64, /*cache_size=*/128,
+                    /*filter_data=*/true, /*filter_instrs=*/k == 1));
+        if (cache_filter->get_error_string() != "") {
+            fprintf(stderr, "Couldn't construct a cache_filter %s",
+                    cache_filter->get_error_string().c_str());
+            return false;
+        }
+        filters.push_back(cache_filter.get());
+
+        auto type_filter =
+            std::unique_ptr<dynamorio::drmemtrace::record_filter_t::record_filter_func_t>(
+                new dynamorio::drmemtrace::type_filter_t(
+                    { TRACE_TYPE_ENCODING, TRACE_TYPE_MARKER },
```

Comment:
But we don't want to remove all markers: looks like a bug.
Update: I see special casing to ignore this: see other comments as this is confusing.

Reply:
Fixed.